File: BOJ/BOJ_19585.py
# BOJ_19585
# 전설

import sys

inp = sys.stdin.readline


class TrieNode:
    def __init__(self):
        self.children = dict()
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root
        for char in word:
            if char not in cur.children:
                cur.children[char] = TrieNode()
            cur = cur.children[char]
        cur.is_end = True

    def search(self, word, nickname_set):
        cur = self.root
        for i in range(min(len(word) - 1, 1000)):
            char = word[i]
            if char not in cur.children:
                return False
            cur = cur.children[char]
            if cur.is_end:
                nickname = word[i + 1 :]
                if nickname in nickname_set:
                    return True
        return False


def main():
    c, n = map(int, inp().split())
    trie = Trie()
    nickname_set = set()

    for _ in range(c):
        trie.insert(inp().strip())
    for _ in range(n):
        nickname_set.add(inp().strip())

    q = int(inp())
    for _ in range(q):
        ret = trie.search(inp().strip(), nickname_set)
        if ret:
            print("Yes")
        else:
            print("No")


if __name__ == "__main__":
    main()

File: BOJ/BOJ_19585.js
// BOJ_19585
// 전설

const fs = require('fs');
const filePath = process.platform === 'linux' ? 0 : './input.txt';
const input = fs.readFileSync(filePath, 'utf-8').trim().split('\n');

class TrieNode {
  constructor() {
    this.children = {};
    this.isEndOfWord = false;
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let curNode = this.root;

    for (const char of word) {
      if (!curNode.children.hasOwnProperty(char)) {
        curNode.children[char] = new TrieNode();
      }
      curNode = curNode.children[char];
    }
    curNode.isEndOfWord = true;
  }

  search(word, nicknameSet) {
    let curNode = this.root;

    for (let i = 0; i < Math.min(word.length - 1, 1000); i++) {
      const char = word[i];
      if (!curNode.children.hasOwnProperty(char)) {
        return false;
      }
      curNode = curNode.children[char];
      if (curNode.isEndOfWord) {
        const nickname = word.slice(i + 1);
        if (nicknameSet.has(nickname)) {
          return true;
        }
      }
    }

    return false;
  }
}

function main(inputArr) {
  const [c, n] = inputArr[0].split(' ').map(Number);
  const trie = new Trie();
  const nicknameSet = new Set();

  let line = 1;

  for (let i = 0; i < c; i++) {
    trie.insert(inputArr[line++]);
  }
  for (let i = 0; i < n; i++) {
    nicknameSet.add(inputArr[line++]);
  }

  const res = [];

  let q = Number(inputArr[line++]);
  while (q--) {
    res.push(trie.search(inputArr[line++], nicknameSet));
  }

  return res
    .map((val) => {
      return val === true ? 'Yes' : 'No';
    })
    .join('\n');
}

console.log(main(input));
